### ISSUE TITLE
optimizing a vector as a field in Style doesn't work

### ISSUE THREAD

**Issue by hypotext** [penrose/penrose#501]
There is no gradient for the very simple program triple that @srossd sent to us today, though there is one constraint, and the generated energy appears to be correct. Maybe an issue with finding `AccessPath`?

energy:
```
(function anonymous(x0,x1,x2,x3,x4
) {
const x5 = 0;
const x6 = 0;
const x7 = 252.1585483546197;
console.log("x1 (var x7) | value: ", x7, "during energy evaluation");
const x8 = 75.30582642063939;
console.log("x2 (var x8) | value: ", x8, "during energy evaluation");
const x9 = x7 - x8;
console.log("energy (var x9) | value: ", x9, "during energy evaluation");
const x10 = 0;
const x11 = Math.max(x9, x10);
const x12 = x11 * x11;
const x13 = x6 + x12;
const x14 = 100000;
const x15 = x14 * x0;
const x16 = x13 * x15;
const x17 = x5 + x16;
return x17;
})
```

grad: (wrong)
```
generated f ƒ anonymous(x0,x1,x2,x3,x4
) {
const x5 = 0;
const x6 = 0;
const x7 = 0;
const x8 = 0;
return [x5, x6, x7, x8];
}
```

Now sure why/how the energy would be correct but not the gradient.

Oddly, a very similar function appears to work in the `graph-domain` branch: https://github.com/penrose/penrose/blob/graph-domain/examples/graph-domain/graph-theory.sty#L110

It appears to be linked to the use of fields, or perhaps path aliases/references in general. This Style program works when `o.v1` and `o.v2` are removed, and it doesn't seem to matter whether `start` and `end` are declared explicitly as varying `vec2s` or not.

```
Object o {
    -- Remove these lines
    -- vec2 o.v1 = (?, ?)
    -- vec2 o.v2 = (?, ?)

    shape o.shape = Line {
        -- Optional: start: (?, ?) -- remove this: o.v1
        -- Optional: end: (?, ?) -- remove this: o.v2
        color: rgba(0., 0., 0., 1.)
        thickness: 5.
        rightArrowhead: true
    }

    ensure rightwards(o.shape) -- ensure lines move from left to right
}
```

**Comment by hypotext** [comment]
One of the problems here is that vector types are not expressive enough: `Vector` can only hold `Expr`s (i.e. no `VarAD`s for autodiff); `VectorV` can only hold `Value`s. So, something like `(?, x.c + y.v)` could not be modeled in either type in the translation. 

Currently, if a field is set to a vector, all elements are auto-converted to `Expr`s, losing autodiff info. That's why the gradients are not present. (Though not clear why some of the energy info is there...)

One (kind of messy) solution is to store `VarAD`s only in `varyingMap`, and always look them up there first. I did try that--not clear why that solution didn't fully work. Will need to investigate this more.

**Comment by hypotext** [comment]
Proposed solution: Change all container types to look like this. `(Expr | Value<VarAD>, Expr | Value<VarAD>)`

**Comment by hypotext** [comment]
Weirdly, the non-working minimal example appears to work now (run 589779a). But individual accesspath operations may still be a problem.

**Comment by hypotext** [comment]
Previous comment was not true as I was running the wrong minimal case. Now resolved for real in 14215184d7c40da713fb677f813bbdd49a8cc444.
